X11/clients/xclipboard/xclipboard.man
.TH XCLIPBOARD 1 "Release 4" "X Version 11"
.de Ds
.nf
.\\$1D \\$2 \\$1
.ft 1
.ps \\n(PS
.\".if \\n(VS>=40 .vs \\n(VSu
.\".if \\n(VS<=39 .vs \\n(VSp
..
.de De
.ce 0
.if \\n(BD .DF
.nr BD 0
.in \\n(OIu
.if \\n(TM .ls 2
.sp \\n(DDu
.fi
..
.SH NAME
xclipboard - X clipboard client
.SH SYNOPSIS
\fBxclipboard\fP [ \fI-toolkitoption\fP ...] [-w] [-nw]
.SH DESCRIPTION
The \fIxclipboard\fP program is used to collect and display text selections 
that are sent to the CLIPBOARD by other clients.  It is typically used to
save CLIPBOARD selections for later use.  It stores each CLIPBOARD selection
as a separate string, each of which can be selected.  Each time CLIPBOARD
is asserted by another application, \fIxclipboard\fP transfers the contents
of that selection to a new buffer and displays it in the text window.
Buffers are never automatically deleted, so you'll want to use the delete
button to get rid of useless items.
.PP
Since \fIxclipboard\fP uses a Text Widget to display the contents of the
clipboard, text sent to the CLIPBOARD may be re-selected for use in other
applications.  \fIxclipboard\fP also responds to requests for the CLIPBOARD
selection from other clients by sending the entire contents of the currently
displayed buffer.
.PP
An \fIxclipboard\fP window has the following buttons across the top:
.TP 8
.I quit
When this button is pressed, \fIxclipboard\fP exits.
.TP 8
.I delete
When this button is pressed, the current buffer is deleted and the
next one displayed.
.TP 8
.I new
Creates a new buffer with no contents.  Useful in constructing a new
CLIPBOARD selection by hand.
.TP 8
.I next
Displays the next buffer in the list.
.TP 8
.I previous
Displays the previous buffer.
.SH OPTIONS
The \fIxclipboard\fP program accepts all of the standard X Toolkit command 
line options as well as the following:
.TP 8
.B \-w
This option indicates that lines of text that are too long to be displayed on
one line in the clipboard should wrap around to the following lines.
.TP 8
.B \-nw
This option indicates that long lines of text should not wrap around.  This
is the default behavior.
.SH WIDGETS
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxclipboard\fR.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp
.nf
.TA .5i 1.0i 1.5i
.ta .5i 1.0i 1.5i
XClipboard  xclipboard
	Form  form
		Command  quit
		Command  delete
		Command  new
		Command  next
		Command  prev
		Text  text
.fi
.sp
.SH SENDING/RETRIEVING CLIPBOARD CONTENTS
Text is copied to the clipboard whenever a client asserts ownership of the
\fBCLIPBOARD\fP selection.  Text is copied from the clipboard whenever a
client requests the contents of the \fBCLIPBOARD\fP selection.  Examples of
event bindings that a user may wish to include in a resource configuration
file to use the clipboard are:
.sp
.Ds 5
.TA .5i
.ta .5i 3.0i
*VT100.Translations: #override \\
	<Btn3Up>:	select-end(CLIPBOARD) \\n\\
	<Btn2Up>:	insert-selection(PRIMARY,CLIPBOARD) \\n\\
	<Btn2Down>:	ignore ()

.De
.sp
.SH "SEE ALSO"
X(1), xcutsel(1), xterm(1), individual client documentation for how to make a 
selection and send it to the CLIPBOARD.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH FILES
/usr/lib/X11/app-defaults/XClipboard - specifies required resources
.SH COPYRIGHT
Copyright 1988, Massachusetts Institute of Technology
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHOR
Ralph R. Swick, DEC/MIT Project Athena
.br
Chris D. Peterson, MIT X Consortium
.br
Keith Packard, MIT X Consortium


X11/CHANGES
The following changes have been made to this directory since R3:

     o	the configuration files have moved from util/imake.includes/ to config/
     o	sources for imake are now in config/ instead of util/imake/.



X11/clients/xclipboard/xclipboard.c
/*
 * $XConsortium: xclipboard.c,v 1.15 89/12/15 20:11:09 keith Exp $
 *
 * Copyright 1989 Massachusetts Institute of Technology
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in advertising
 * or publicity pertaining to distribution of the software without specific,
 * written prior permission.  M.I.T. makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * M.I.T. DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL M.I.T.
 * BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
 * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Author:  Ralph Swick, DEC/Project Athena
 * Updated for R4:  Chris D. Peterson,  MIT X Consortium.
 * Reauthored by: Keith Packard, MIT X Consortium.
 */

/* $XConsortium: xclipboard.c,v 1.15 89/12/15 20:11:09 keith Exp $ */

#include <stdio.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xatom.h>

#include <X11/Xmu/Atoms.h>
#include <X11/Xmu/StdSel.h>

#include <X11/Xaw/Form.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xaw/Cardinals.h>

#define Command commandWidgetClass
#define Text    asciiTextWidgetClass

#define INFINITY 10000000	/* pretty big, huh? */

typedef struct _Clip {
    struct _Clip    *next, *prev;
    char	    *clip;
    int		    avail;
} ClipRec, *ClipPtr;

extern char *malloc ();

static long TextLength (w)
    Widget  w;
{
    return XawTextSourceScan (XawTextGetSource (w),
			      (XawTextPosition) 0,
 			      XawstAll, XawsdRight, 1, TRUE);
}

SaveClip (w, clip)
    Widget  w;
    ClipPtr clip;
{
    Arg	    args[1];
    char    *data;
    int	    len;
    Widget  source;

    source = XawTextGetSource (w);
    XtSetArg (args[0], XtNstring, &data);
    XtGetValues (source, args, 1);
    len = strlen (data);
    if (len >= clip->avail)
    {
	if (clip->clip)
	    free (clip->clip);
	clip->clip = malloc (len + 1);
	if (!clip->clip)
	    clip->avail = 0;
	else
	    clip->avail = len + 1;
    }
    if (clip->avail)
    {
	strcpy (clip->clip, data);
    }
}

RestoreClip (w, clip)
    Widget  w;
    ClipPtr clip;
{
    Arg	    args[1];
    Widget  source;

    source = XawTextGetSource (w);
    XtSetArg (args[0], XtNstring, clip->clip);
    XtSetValues (source, args, 1);
}

ClipPtr
NewClip (w, old)
    Widget  w;
    ClipPtr old;
{
    ClipPtr newClip;

    newClip = (ClipPtr) malloc (sizeof (ClipRec));
    if (!newClip)
	return newClip;
    newClip->clip = 0;
    newClip->avail = 0;
    newClip->prev = old;
    newClip->next = NULL;
    if (old)
    {
	newClip->next = old->next;
	old->next = newClip;
    }
    return newClip;
}

DeleteClip (w, clip)
    Widget  w;
    ClipPtr clip;
{
    if (clip->prev)
	clip->prev->next = clip->next;
    if (clip->next)
	clip->next->prev = clip->prev;
    if (clip->clip)
	free (clip->clip);
    free ((char *) clip);
}

static ClipPtr	currentClip;
static Widget	text, nextButton, prevButton;

static void set_button_state ()
{
    Boolean prevvalid, nextvalid;
    Arg arg;

    prevvalid = False;
    nextvalid = False;
    if (currentClip)
    {
	prevvalid = currentClip->prev != NULL;
	nextvalid = currentClip->next != NULL;
    }
    arg.name = XtNsensitive;
    arg.value = (XtArgVal) prevvalid;
    XtSetValues (prevButton, &arg, ONE);
    arg.value = (XtArgVal) nextvalid;
    XtSetValues (nextButton, &arg, ONE);
}

static void
NextCurrentClip ()
{
    if (currentClip && currentClip->next)
    {
	SaveClip (text, currentClip);
	currentClip = currentClip->next;
	RestoreClip (text, currentClip);
	set_button_state ();
    }
}

static void
PrevCurrentClip ()
{
    if (currentClip && currentClip->prev)
    {
	SaveClip (text, currentClip);
	currentClip = currentClip->prev;
	RestoreClip (text, currentClip);
	set_button_state ();
    }
}

static void
DeleteCurrentClip ()
{
    ClipPtr newCurrent;
    if (currentClip)
    {
	if (currentClip->prev)
	{
	    newCurrent = currentClip->prev;
	}
	else
	{
	    newCurrent = currentClip->next;
	}
	DeleteClip (text, currentClip);
	currentClip = newCurrent;
	if (currentClip)
	    RestoreClip (text, currentClip);
	else
	{
	    EraseTextWidget ();
	}
	set_button_state ();
    }
}

static void
Quit ()
{
    XtCloseDisplay  (XtDisplay (text));
    exit (0);
}

static void
NewCurrentClip ()
{
    NewCurrentClipContents ("", 0);
}

NewCurrentClipContents (data, len)
    char    *data;
    int	    len;
{
    XawTextBlock textBlock;
    ClipPtr newCurrent;

    if (!currentClip && TextLength (text))
	currentClip = NewClip (text, (ClipPtr) 0);
    if (currentClip)
	SaveClip (text, currentClip);
    /* append new clips at the end */
    while (currentClip && currentClip->next)
	currentClip = currentClip->next;
    newCurrent = NewClip (text, currentClip);
    
    currentClip = newCurrent;

    textBlock.ptr = data;
    textBlock.firstPos = 0;
    textBlock.length = len;
    textBlock.format = FMT8BIT;
    if (XawTextReplace(text, 0, TextLength (text), &textBlock))
	XBell( XtDisplay(text), 0);
    set_button_state ();
}

EraseTextWidget ()
{
    XawTextBlock block;

    block.ptr = "";
    block.length = 0;
    block.firstPos = 0;
    block.format = FMT8BIT;

    XawTextReplace(text, 0, INFINITY, &block);
    /* If this fails, too bad. */
}


XtActionsRec xclipboard_actions[] = {
    "NewClip", NewCurrentClip,
    "NextClip",	NextCurrentClip,
    "PrevClip", PrevCurrentClip,
    "DeleteClip", DeleteCurrentClip,
    "Quit", Quit,
};

static XrmOptionDescRec table[] = {
    {"-w",	    "*text*wrap",		XrmoptionNoArg,  "Word"},
    {"-nw",	    "*text*wrap",		XrmoptionNoArg,  "Never"},
};

static void	LoseSelection ();
static void	InsertClipboard ();
static Boolean	ConvertSelection();

static void 
InsertClipboard(w, client_data, selection, type, value, length, format)
Widget w;
caddr_t client_data;
Atom *selection, *type;
caddr_t value;
unsigned long *length;
int *format;
{
    if (*type == 0 /*XT_CONVERT_FAIL*/ || *length == 0) {
	XBell( XtDisplay(w), 0 );
	return;
    }
    
    NewCurrentClipContents ((char *) value, *length);

    XtOwnSelection(w, XA_CLIPBOARD(XtDisplay(w)), CurrentTime,
		   ConvertSelection, LoseSelection, NULL);

    XtFree(value);
}

static Boolean ConvertSelection(w, selection, target,
				type, value, length, format)
    Widget w;
    Atom *selection, *target, *type;
    caddr_t *value;
    unsigned long *length;
    int *format;
{
    Display* d = XtDisplay(w);
    XSelectionRequestEvent* req =
	XtGetSelectionRequest(w, *selection, (XtRequestId)NULL);

    if (*target == XA_TARGETS(d)) {
	Atom* targetP;
	Atom* std_targets;
	unsigned long std_length;
	XmuConvertStandardSelection(w, req->time, selection, target, type,
				  (caddr_t*)&std_targets, &std_length, format);
	*value = XtMalloc(sizeof(Atom)*(std_length + 5));
	targetP = *(Atom**)value;
	*targetP++ = XA_STRING;
	*targetP++ = XA_TEXT(d);
	*targetP++ = XA_LENGTH(d);
	*targetP++ = XA_LIST_LENGTH(d);
	*targetP++ = XA_CHARACTER_POSITION(d);
	*length = std_length + (targetP - (*(Atom **) value));
	bcopy((char*)std_targets, (char*)targetP, sizeof(Atom)*std_length);
	XtFree((char*)std_targets);
	*type = XA_ATOM;
	*format = 32;
	return True;
    }

    if (*target == XA_LIST_LENGTH(d) ||
	*target == XA_LENGTH(d))
    {
    	long * temp;
    	
    	temp = (long *) XtMalloc(sizeof(long));
    	if (*target == XA_LIST_LENGTH(d))
      	  *temp = 1L;
    	else			/* *target == XA_LENGTH(d) */
      	  *temp = (long) TextLength (w);
    	
    	*value = (caddr_t) temp;
    	*type = XA_INTEGER;
    	*length = 1L;
    	*format = 32;
    	return True;
    }
    
    if (*target == XA_CHARACTER_POSITION(d))
    {
    	long * temp;
    	
    	temp = (long *) XtMalloc(2 * sizeof(long));
    	temp[0] = (long) 0;
    	temp[1] = TextLength (w);
    	*value = (caddr_t) temp;
    	*type = XA_SPAN(d);
    	*length = 2L;
    	*format = 32;
    	return True;
    }
    
    if (*target == XA_STRING ||
      *target == XA_TEXT(d) ||
      *target == XA_COMPOUND_TEXT(d))
    {
	extern char *_XawTextGetSTRING();
    	if (*target == XA_COMPOUND_TEXT(d))
	    *type = *target;
    	else
	    *type = XA_STRING;
	*length = TextLength (w);
    	*value = _XawTextGetSTRING((TextWidget) w, 0, *length);
    	*format = 8;
    	return True;
    }
    
    if (XmuConvertStandardSelection(w, req->time, selection, target, type,
				    value, length, format))
	return True;

    return False;
}

static void LoseSelection(w, selection)
    Widget w;
    Atom *selection;
{
    XtGetSelectionValue(w, *selection, XA_STRING, InsertClipboard,
			NULL, CurrentTime);
}

/*ARGSUSED*/
static Boolean RefuseSelection(w, selection, target,
			       type, value, length, format)
    Widget w;
    Atom *selection, *target, *type;
    caddr_t *value;
    unsigned long *length;
    int *format;
{
    return False;
}

/*ARGSUSED*/
static void LoseManager(w, selection)
    Widget w;
    Atom *selection;
{
    XtError("another clipboard has taken over control\n");
}

void
main(argc, argv)
int argc;
char **argv;
{
    Arg args[2];
    Widget top, parent, quit, delete, new;
    Atom manager;

    top = XtInitialize( "xclipboard", "XClipboard", table, XtNumber(table),
			  &argc, argv);

    XtAddActions (xclipboard_actions, XtNumber (xclipboard_actions));
    /* CLIPBOARD_MANAGER is a non-standard mechanism */
    manager = XInternAtom(XtDisplay(top), "CLIPBOARD_MANAGER", False);
    if (XGetSelectionOwner(XtDisplay(top), manager))
	XtError("another clipboard is already running\n");

    parent = XtCreateManagedWidget("form", formWidgetClass, top, NULL, ZERO);
    quit = XtCreateManagedWidget("quit", Command, parent, NULL, ZERO);
    delete = XtCreateManagedWidget("delete", Command, parent, NULL, ZERO);
    new = XtCreateManagedWidget("new", Command, parent, NULL, ZERO);
    nextButton = XtCreateManagedWidget("next", Command, parent, NULL, ZERO);
    prevButton = XtCreateManagedWidget("prev", Command, parent, NULL, ZERO);

    XtSetArg(args[0], XtNtype, XawAsciiString);
    XtSetArg(args[1], XtNeditType, XawtextEdit);
    text = XtCreateManagedWidget( "text", Text, parent, args, TWO);

    set_button_state ();

    XtRealizeWidget(top);

    XtOwnSelection(text, manager, CurrentTime,
		   RefuseSelection, LoseManager, NULL);
    XtOwnSelection(text, XA_CLIPBOARD(XtDisplay(text)), CurrentTime,
		   ConvertSelection, LoseSelection, NULL);

    XtMainLoop();
}
